### PR TITLE
Add release compilation from []IndexRelease & []Bundle

### PR DESCRIPTION
**Pull request by tuommaki**
To replace earlier release aggregation, implement release compilation
based on defined release index and available version bundles.

CompileReleases() creates canonicalized collection of releases.

Towards https://github.com/giantswarm/giantswarm/issues/3047

### CODE REVIEW

File: index_release.go [hunk 1]
```diff
@@ -19,6 +19,51 @@ type IndexRelease struct {
...
+		release := Release{
+			active:    ir.Active,
+			timestamp: ir.Date.Format(releaseTimestampFormat),
+			version:   ir.Version,
+		}
+
+		for _, a := range ir.Authorities {
+			b, found := bundleCache[a.BundleID()]
+			if !found {
+				return nil, microerror.Maskf(invalidReleaseError, "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles.")
```
[the remark this answers is not in the document]
Comment (tuommaki) [reply index_release.go]: Ok. Makes sense. I'll change this.

File: index_release.go [hunk 2]
```diff
@@ -19,6 +20,95 @@ type IndexRelease struct {
...
+
+	type LogState int
+	const (
+		New LogState = iota
+		Removed
+	)
+
+	sort.Sort(SortReleasesByVersion(releases))
+
+	prevChangelogs := make(map[string]LogState)
```
Comment (tuommaki) [inline index_release.go]: This map with `LogState` might be unnecessary. I don't fully use it yet, but this was my first intuition.

File: index_release.go [hunk 3]
```diff
@@ -19,6 +20,95 @@ type IndexRelease struct {
...
+		}
+
+		for _, a := range ir.Authorities {
+			b, found := bundleCache[a.BundleID()]
+			if !found {
+				logger.Log("level", "warning", "message", "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles. Skipping.")
+				continue INDEX_RELEASES
+			}
+			release.bundles = append(release.bundles, b)
+		}
```
Comment (xh3b4sd) [inline index_release.go]: Can we have some function `groupBundlesForIndexRelease(ir, bundles)`?  It can return the list of grouped bundles and a not found error in case some bundle is missing. That way we get rid of the loop label. I hate these. :D IMO the code would be easier to understand then and we can assign the returned list or do the logging separated from the grouping logic.

**Comment by tuommaki** [comment]
> The changelog deduplication code looks a bit complicated but I guess this is for reasons.

Yes and no. It's partially just my first untested intuition, but somewhat also because of complex nature. I just pushed test case for introduction of patch release and it doesn't pass, so I have a problem to solve for tomorrow here. Good to have a test case though so I can work with the code. It's highly probable that I need to rethink the code and throw away the current one.

**Comment by tuommaki** [comment]
Ok. I feel this is good for a review. I didn't write that much tests for release compilation so if you feel it deserves more, please write a note.

*IMPORTANT*: :warning: 
Please review changelog deduplication tests carefully and give me a note if you have any kind of corner case in mind so I can add tests and therefore more verification for deduplication algorithm. I feel personally that one of the big challenges was newly introduced patch release which is now covered.

File: index_release.go [hunk 4]
```diff
@@ -19,6 +20,152 @@ type IndexRelease struct {
...
+	for _, b := range bundles {
+		bundleCache[b.ID()] = b
+	}
+
+	var releases []Release
+
+	for _, ir := range indexReleases {
+		bundles, err := groupBundlesForIndexRelease(ir, bundleCache)
+		if IsBundleNotFound(err) {
+			logger.Log("level", "warning", "message", err.Error())
```
Comment (xh3b4sd) [inline index_release.go]: Super nitpick, but I would like to have it like this to be more explicit. Below is also logging that we can align with this approach then. 

```
logger.Log("level", "warning", "message", "failed grouping version bundles for release", "stack", fmt.Sprintf("%#v", err))
```

File: index_release.go [hunk 5]
```diff
@@ -19,6 +20,152 @@ type IndexRelease struct {
...
+			continue
+		}
+
+		if err != nil {
+			panic(err)
```
Comment (xh3b4sd) [inline index_release.go]: Lets please not panic. We should rather handle the error and return it to the caller so he can decide what to do with it. Just as a general rule. Paniccing at runtime should be the exception. We can do better here.

Comment (tuommaki) [reply index_release.go]: In general I'd have liked to panic if release compilation fails with available data to any other error than missing bundle, but now that you mentioned it, this is actually very wrong thing to do in library. Changed this. Thanks.

File: index_release.go [hunk 6]
```diff
@@ -19,6 +21,155 @@ type IndexRelease struct {
...
+			return nil, microerror.Maskf(bundleNotFoundError, "IndexRelease v%s contains Authority with bundle ID %s that cannot be found from collected version bundles.", ir.Version, a.BundleID())
+		}
+		groupedBundles = append(groupedBundles, b)
+	}
+
+	return groupedBundles, nil
+}
+
+// deduplicateReleaseChangelog removes duplicate changelog entries in
+// consequtive release entries. Core concept of algorithm here is to first sort
```
Comment (MarcelMue) [inline index_release.go]: `consecutive`

File: index_release.go [hunk 7]
```diff
@@ -19,6 +21,155 @@ type IndexRelease struct {
...
+	return filteredReleases
+}
+
+// findPreviousRelease finds release that is older than argument r0. This
+// function expects that releases is sorted by version as it is iterated
+// backwards. If no previous release is found, empty Release is returned.
+func findPreviousRelease(r0 Release, releases []Release) Release {
+	r0Date, err := time.Parse(releaseTimestampFormat, r0.Timestamp())
+	if err != nil {
+		panic(err)
```
Comment (MarcelMue) [inline index_release.go]: Why are there still so many panics? I don't understand where we are using this, that makes panics okay.

File: index_release_test.go [hunk 8]
```diff
@@ -17,6 +22,1049 @@ func urlMustParse(v string) *URL {
...
+						{
+							Component:   "bar-operator",
+							Description: "changed feature y",
+							Kind:        KindChanged,
+						},
+					},
+					version: "2.0.1",
+				},
+			},
+		},
```
Comment (MarcelMue) [inline index_release_test.go]: I thougth of a case where I am not sure how we handle this or if we allow this to happen:
 - version `1.0.0` is released
 -  version `1.1.0` is released
 - now patch `1.0.1` is released

Can this happen in this timeline? Or would we always force the patch to be `1.1.1` ?

Comment (xh3b4sd) [reply index_release_test.go]: This is what the last test case is about. At least I understood it that way. Do you mean the same or where is the difference?

Comment (MarcelMue) [reply index_release_test.go]: Oh you are right, I misread the last testcase and didn't look at the timestamp closely enough.

Ignore this comment.

File: index_release.go (hunk 7 above)
Comment (tuommaki) [reply index_release.go]: Panics here came from my thinking that in this case if parsing fails, it means timestamps are formatted wrong, which means `opsctl` validation broke and if we just pass error and later in the call chain log and ignore it, we don't necessarily notice the issue. When cluster-service panics, we definitely notice the issue but we still don't affect customer workloads.

I can change these to errors and there is a very strong point that library should never panic, but how can we make sure we make this case visible enough in cluster-service? Should we expose a metric of failed releases (I mean number of IndexReleases that couldn't be compiled to Release)?

Comment (MarcelMue) [reply index_release.go]: I think a metric for errors in cluster service should do the job. Paging oncall that way lets us add information to the alert which lets them react fast to what's happening.
The service panicking doesn't give oncall a lot information right away IMO, especially if it restarts and then starts crashlooping.

Comment (tuommaki) [reply index_release.go]: > I think a metric for errors in cluster service should do the job. Paging oncall that way lets us add information to the alert which lets them react fast to what's happening.
The service panicking doesn't give oncall a lot of information right away IMO, especially if it restarts and then starts crashlooping.

Good point. It returns error now :)

**Comment by tuommaki** [comment]
I feel this is pretty good now. Review comments are sorted and tests seem good. Merging this PR. If there are any concerns / thoughts coming up later, lets get them fixed in a new PR.
